[17기 박재현] 상품 리뷰 조회, 정렬 및 필터링

namujinju opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 특정 상품에 대한 리뷰 조회, 정렬 및 필터링
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 특정 상품에 대한 모든 리뷰 조회
- 리뷰 생성 시간, 좋아요 순에 따른 정렬
- 여러 개의 별점 조건에 따른 필터링

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

namujinju committed:
Add: 상품 리뷰 조회 view 생성
- 특정 상품에 대한 모든 리뷰 조회
- 리뷰 생성 시간, 좋아요 순에 따른 정렬
- 별점에 따른 필터링

soheon-lee commented on product/urls.py:
```diff
@@ -0,0 +1,9 @@
+from product.views import (
+    ProductReviewView
+)
+
+urlpatterns = [
+    path('/<int:product_id>/review', ProductReviewView.as_view()),
+]
```
RESTful URL goooood!

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,71 @@
+from django.views     import View
+from django.db.models import Q
+from django.db.utils  import DataError
+
+from product.models import *
+
```
import * 을 하면 내부에 있는 모든 객체를 한 번에 가져오게 됩니다. 모든 객체를 다 사용하지 않을 가능성도 있으므로 사용하고 있는 개별 객체를 하나하나 import 해주세요 :)

namujinju replied:
네 알겠습니다. 수정하겠습니다!

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,71 @@
+    def get(self, request, product_id):
+        try:
+            product   = Product.objects.get(id=product_id)
+            order     = request.GET.get('order', None)
+            rate_list = request.GET.getlist('rate', None)
+            like      = request.GET.get('like', None)
```
getlist 사용하셨군요!! good

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,71 @@
+            
+            if not product_reviews:
+                return JsonResponse({'message':'NO_REVIEWS'}, status=200)
+
+            results = []
+
+            for product_review in product_reviews:
+                results.append(
+                    {
+                        "review_id":product_review.id,
+                        "review_content":product_review.content,
+                        "review_image":product_review.image_url,
+                        "review_rate":product_review.rate,
+                        "product_name":product.name,
+                        "created_at":product_review.created_at,
+                        "review_user_name":product_review.user.name,
+                        "review_like":ReviewLike.objects.filter(review=product_review).count(),
+                    }
+                )
+
```
이 부분 `list comprehension` 으로 한 번 수정해보세요 :)

namujinju replied:
네 알겠습니다

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,71 @@
+            for product_review in product_reviews:
+                results.append(
+                    {
+                        "review_id":product_review.id,
+                        "review_content":product_review.content,
+                        "review_image":product_review.image_url,
+                        "review_rate":product_review.rate,
+                        "product_name":product.name,
+                        "created_at":product_review.created_at,
+                        "review_user_name":product_review.user.name,
+                        "review_like":ReviewLike.objects.filter(review=product_review).count(),
```
`:` 로 코드 정렬해주세요

namujinju replied:
네 수정하겠습니다

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,71 @@
+        
+        except KeyError:
+            return JsonResponse({'message':'KEY_ERROR'}, status=400)
+        
+        except DataError:
+            return JsonResponse({'message':'DATA_ERROR'}, status=400)
+
```
이 DataError은 어떤 경우에 발생하나요?

namujinju replied:
DataError는 테이블에 데이터를 집어넣을 때 그 데이터가 필드 조건(max_length 등)에 맞지 않을 때 발생합니다.
상품 리뷰를 보여주기만 하는 이 뷰에선 필요가 없을 것 같습니다.
다른 뷰에서 코드를 복붙해올 때 예외 처리를 꼼꼼하게 살펴보지 않은 것 같습니다. 수정하겠습니다!

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,71 @@
+                q = Q(product=product)
+            
+            product_reviews = ProductReview.objects.filter(q)
+            
+            if not product_reviews:
+                return JsonResponse({'message':'NO_REVIEWS'}, status=200)
+
```
'NO_REVIEWS'인데 status_code 200OK 인가요?

namujinju replied:
상품에 리뷰가 없는 경우라 에러는 아니기 때문에 status_code를 200으로 했습니다.
리뷰가 없는 경우 빈 리스트만 보내줘도 되지만 json response로 프론트에 리뷰가 없다고 알려주기 위해 작성했습니다.
지금 생각해보니 프론트에서도 상품 리뷰가 없는 경우를 구분할 수 있으니  굳이 메시지를 보낼 필요가 없었던 것 같습니다.
프론트와도 한 번 상의 후 필요 없으면 지우겠습니다.

soheon-lee replied:
네!

soheon-lee commented on product/views.py:
```diff
@@ -1,3 +1,71 @@
+                'recent':'created_at',
+                'like':'review_like',
+            }
+            
+            if order in order_dict:
+                results.sort(key=lambda result: result[order_dict[order]], reverse=True)
+
```
파이썬 sort 메소드를 사용하셨군요! :)
django에도 데이터를 정렬할 수 있는 메소드가 있습니다 :) 한 번 찾아서 적용해보시고,
`python sort()`와 어떻게 다른지 공부해보시는 것을 추천드립니다.

namujinju replied:
네. 다른 팀원도 같은 피드백을 받았는데 저도 미리 공부해서 적용해볼 걸 그랬네요. 공부 후 수정하겠습니다!

namujinju replied:
order_by는 알고 있었지만 annotate 개념이 생소해서 많이 어려웠습니다. 이해하고 나니 굉장히 좋은 기능인 것 같고 앞으로도 활용할 일이 많을 것 같습니다. 감사합니다!

namujinju committed:
Refactor: product view 코드 삭제 및 정렬
- results에 결과를 넣을 때 append에서 list comprehension으로 수정
- 불필요한 예외 처리 삭제
- 모델 클래스에서 인스턴스를 정렬해서 불러오도록 수정

namujinju committed:
Modify: 불필요 데이터 제거
- 정렬을 위해 쓰였던 results의 created_at 제거

soheon-lee commented:
LGTM! 재현님 수고하셨습니당!! Keep working 💪
